### PR TITLE
Challenge

### PR DESCRIPTION
**Pull request by xiaoZ857**

### CODE REVIEW

File: contracts/Campaign.sol [hunk 1]
```diff
@@ -166,6 +170,115 @@ contract Campaign is ICampaign, Ownable, ERC721 {
...
+      uint256 _tranReward = properties[_cheaterId].pendingReward;
+      properties[_cheaterId].pendingReward = 0;
+      properties[challengeRecords[challengeRecordId].challengerId].pendingReward += _tranReward * 3/5;
+      sharedReward += _tranReward * 3/10;
+      IERC20(targetToken).safeTransfer(Consts.PROTOCOL_RECIPIENT, _tranReward/10);
+
+      emit EvFailure(_cheaterId);
+    } else{
+      cheatCount += 1;
+
```
Comment (noyyyy) [inline contracts/Campaign.sol:268]: why cheatCount += 1

**Review by noyyyy — CHANGES_REQUESTED** [review]

**Comment by noyyyy** [comment]
run `git pull origin main --rebase` to fetch the newest change from main.

File: contracts/Campaign.sol [hunk 2]
```diff
@@ -166,6 +170,115 @@ contract Campaign is ICampaign, Ownable, ERC721 {
...
+    external
+    override
+    onlyChallengeEnded(challengeRecordId)
+    onlyChallengeExist(challengeRecordId)
+    onlyNotJudged(challengeRecordId)
+  { 
+    uint256 _cheaterId = challengeRecords[challengeRecordId].cheaterId;
+    uint256 _count = challengeRecords[challengeRecordId].agreeCount + challengeRecords[challengeRecordId].disagreeCount;
+    bool _legal = (_count >= _idx * 2/3);
+    require(_legal == true, 'Challenge: not enough voter');
```
[the remark this answers is not in the document]
Comment (xiaoZ857) [reply contracts/Campaign.sol:248]: solved

File: contracts/Campaign.sol [hunk 3]
```diff
@@ -166,6 +170,115 @@ contract Campaign is ICampaign, Ownable, ERC721 {
...
+    challengeRecords[challengeRecordId].state = true;
+
+    if (_result == true) {
+      properties[_cheaterId].tokenStatus = TokenStatus.FAILED;
+
+      uint256 _tranReward = properties[_cheaterId].pendingReward;
+      properties[_cheaterId].pendingReward = 0;
+      properties[challengeRecords[challengeRecordId].challengerId].pendingReward += _tranReward * 3/5;
+      sharedReward += _tranReward * 3/10;
+      IERC20(targetToken).safeTransfer(Consts.PROTOCOL_RECIPIENT, _tranReward/10);
```
[the remark this answers is not in the document]
Comment (xiaoZ857) [reply contracts/Campaign.sol:263]: solved

File: contracts/Campaign.sol [hunk 4]
```diff
@@ -166,6 +170,115 @@ contract Campaign is ICampaign, Ownable, ERC721 {
...
+
+      emit EvFailure(_cheaterId);
+    } else{
+      cheatCount += 1;
+
+      uint256 _tranReward = (properties[_cheaterId].pendingReward) * 2/5;
+      properties[challengeRecords[challengeRecordId].challengerId].pendingReward = (properties[challengeRecords[challengeRecordId].challengerId].pendingReward) * 3/5;
+      sharedReward += _tranReward * 3/4;
+      IERC20(targetToken).safeTransfer(Consts.PROTOCOL_RECIPIENT, _tranReward * 1/4);
+    }
```
[the remark this answers is not in the document]
Comment (xiaoZ857) [reply contracts/Campaign.sol:273]: solved

File: contracts/Campaign.sol [hunk 5]
```diff
@@ -166,6 +170,115 @@ contract Campaign is ICampaign, Ownable, ERC721 {
...
+    if (_result == true) {
+      properties[_cheaterId].tokenStatus = TokenStatus.FAILED;
+
+      uint256 _tranReward = properties[_cheaterId].pendingReward;
+      properties[_cheaterId].pendingReward = 0;
+      properties[challengeRecords[challengeRecordId].challengerId].pendingReward += _tranReward * 3/5;
+      sharedReward += _tranReward * 3/10;
+      IERC20(targetToken).safeTransfer(Consts.PROTOCOL_RECIPIENT, _tranReward/10);
+
+      emit EvFailure(_cheaterId);
```
[the remark this answers is not in the document]
Comment (xiaoZ857) [reply contracts/Campaign.sol:265]: solved by adding a event of cheatment

File: contracts/Campaign.sol (hunk 1 above)
Comment (xiaoZ857) [reply contracts/Campaign.sol:268]: solved

**Pull request merged**